File: go/snake.go
package main

import rl "github.com/gen2brain/raylib-go/raylib"

const (
	GRID_SQUARE_SZ = 20
	GRID_W         = 65
	GRID_H         = 49
	SNAKE_MAX      = GRID_W * GRID_H
	SNAKE_SPEED    = 12.0
)

func main() {
	rl.InitWindow(GRID_W*GRID_SQUARE_SZ, GRID_H*GRID_SQUARE_SZ, "Snake")
	defer rl.CloseWindow()

	var g SnakeGame
	g.Init()

	for !rl.WindowShouldClose() {
		g.Update()
		g.Draw()
	}
}

type SnakeGame struct {
	snake            []rl.Vector2
	snake_direction  Direction
	snake_is_dead    bool
	apple            rl.Vector2
	time_last_update float64
}

var snake_game SnakeGame

func (g SnakeGame) Head() int {
	return len(g.snake) - 1
}

type Direction int

const (
	Direction_Left  Direction = rl.KeyLeft
	Direction_Right Direction = rl.KeyRight
	Direction_Up    Direction = rl.KeyUp
	Direction_Down  Direction = rl.KeyDown
)

func (g *SnakeGame) Init() {
	g.snake = []rl.Vector2{
		{X: GRID_W / 2, Y: GRID_H / 2},
		{X: GRID_W/2 + 1, Y: GRID_H / 2},
		{X: GRID_W/2 + 2, Y: GRID_H / 2},
	}
	g.snake_direction = Direction_Right
	g.snake_is_dead = false
	g.apple = g.RandomVacantCell()
	g.time_last_update = rl.GetTime()
}

func (g *SnakeGame) Update() {
	if g.snake_is_dead {
		if rl.IsKeyPressed(rl.KeySpace) {
			g.Init()
		}
		return
	}

	var pressed_direction Direction
	for key := rl.GetKeyPressed(); key != 0; key = rl.GetKeyPressed() {
		pressed_direction = Direction(key)
		switch {
		case pressed_direction == Direction_Left && g.snake_direction != Direction_Right:
			g.snake_direction = pressed_direction
		case pressed_direction == Direction_Right && g.snake_direction != Direction_Left:
			g.snake_direction = pressed_direction
		case pressed_direction == Direction_Up && g.snake_direction != Direction_Down:
			g.snake_direction = pressed_direction
		case pressed_direction == Direction_Down && g.snake_direction != Direction_Up:
			g.snake_direction = pressed_direction
		}
	}

	if time_now := rl.GetTime(); (time_now-g.time_last_update)*SNAKE_SPEED >= 1 || pressed_direction != 0 {
		g.time_last_update = time_now
		g.UpdateHead()
	}
}

func (g SnakeGame) Draw() {
	rl.BeginDrawing()
	defer rl.EndDrawing()
	rl.ClearBackground(rl.Black)

	if g.snake_is_dead {
		width := rl.MeasureText("GAME OVER", 20)
		rl.DrawText("GAME OVER", (int32(rl.GetScreenWidth())-width)/2, 40, 20, rl.RayWhite)
	}

	for _, cell := range g.snake {
		rl.DrawRectangleRec(CellScreenRec(cell), rl.Lime)
	}
	rl.DrawRectangleRec(CellScreenRec(g.snake[g.Head()]), rl.Green)

	rl.DrawRectangleRec(CellScreenRec(g.apple), rl.Red)
}

func (g SnakeGame) RandomVacantCell() (cell rl.Vector2) {
iteration:
	for range GRID_W * GRID_H {
		cell = rl.Vector2{
			X: float32(rl.GetRandomValue(0, GRID_W-1)),
			Y: float32(rl.GetRandomValue(0, GRID_H-1)),
		}
		for _, snake := range g.snake {
			if rl.Vector2Equals(snake, cell) {
				continue iteration
			}
		}
		break
	}
	return
}

func (g *SnakeGame) UpdateHead() {
	head_new := g.snake[g.Head()]
	switch g.snake_direction {
	case Direction_Left:
		head_new.X--
	case Direction_Right:
		head_new.X++
	case Direction_Up:
		head_new.Y--
	case Direction_Down:
		head_new.Y++
	}

	if CellOutOfBounds(head_new) {
		g.snake_is_dead = true
		return
	}

	for _, body := range g.snake[:g.Head()] {
		if rl.Vector2Equals(body, head_new) {
			g.snake_is_dead = true
			return
		}
	}

	copy(g.snake[:g.Head()], g.snake[1:])
	g.snake[g.Head()] = head_new

	if rl.Vector2Equals(head_new, g.apple) {
		g.snake = append(g.snake, rl.Vector2{})
		copy(g.snake[1:], g.snake[:g.Head()])
		g.apple = g.RandomVacantCell()
	}
}

func CellOutOfBounds(self rl.Vector2) bool {
	return self.X < 0 || self.X >= GRID_W || self.Y < 0 || self.Y >= GRID_H
}

func CellScreenRec(self rl.Vector2) rl.Rectangle {
	return rl.Rectangle{
		X:      self.X * GRID_SQUARE_SZ,
		Y:      self.Y * GRID_SQUARE_SZ,
		Width:  GRID_SQUARE_SZ,
		Height: GRID_SQUARE_SZ,
	}
}

File: c/snake.c
#include <memory.h>
#include <raylib.h>
#include <raymath.h>

#if defined(PLATFORM_WEB)
#include <emscripten/emscripten.h>
#endif

#define GRID_SQUARE_SZ 20
#define GRID_W 65
#define GRID_H 49
#define SNAKE_MAX (GRID_W * GRID_H)
#define SNAKE_SPEED 12.f

bool
CellOutOfBounds(Vector2 cell)
{
	return cell.x < 0 || cell.x >= GRID_W ||
		   cell.y < 0 || cell.y >= GRID_H;
}

Rectangle
CellScreenRec(Vector2 cell)
{
	return (Rectangle){cell.x * GRID_SQUARE_SZ, cell.y * GRID_SQUARE_SZ, GRID_SQUARE_SZ, GRID_SQUARE_SZ};
}

typedef enum Direction
{
	SNAKE_LEFT = KEY_LEFT,
	SNAKE_RIGHT = KEY_RIGHT,
	SNAKE_UP = KEY_UP,
	SNAKE_DOWN = KEY_DOWN,
} Direction;

typedef struct SnakeGame
{
	Vector2 snake[SNAKE_MAX];
	size_t snake_count;
	Direction snake_dir;
	bool snake_is_dead;
	Vector2 apple;
	float time_of_last_update;
} SnakeGame;
SnakeGame gGame = {0};

size_t
SnakeHead(SnakeGame *g)
{
	return g->snake_count - 1;
}

Vector2
RandomVacantCell(SnakeGame *g)
{
	Vector2 cell = {0};
	for (int ITERATION_LIMIT = 0; ITERATION_LIMIT < GRID_W * GRID_H; ITERATION_LIMIT++)
	{
		cell = (Vector2){GetRandomValue(0, GRID_W - 1), GetRandomValue(0, GRID_H - 1)};

		bool hit_snake_body = false;
		for (int i = 0; i < g->snake_count; ++i)
		{
			if (Vector2Equals(g->snake[i], cell))
			{
				hit_snake_body = true;
				break;
			}
		}
		if (hit_snake_body == false)
			return cell;
	}
	return cell;
}

void
InitGame(SnakeGame *g)
{
	g->snake[0] = (Vector2){GRID_W / 2, GRID_H / 2};
	g->snake[1] = (Vector2){GRID_W / 2 + 1, GRID_H / 2};
	g->snake[2] = (Vector2){GRID_W / 2 + 2, GRID_H / 2};
	g->snake_count = 3;

	g->snake_dir = SNAKE_RIGHT;
	g->snake_is_dead = false;
	g->time_of_last_update = GetTime();

	g->apple = RandomVacantCell(g);
}

void
UpdateHead(SnakeGame *g)
{
	Vector2 head_new = g->snake[SnakeHead(g)];
	switch (g->snake_dir)
	{
	case SNAKE_LEFT: head_new.x -= 1; break;
	case SNAKE_RIGHT: head_new.x += 1; break;
	case SNAKE_UP: head_new.y -= 1; break;
	case SNAKE_DOWN: head_new.y += 1; break;
	}

	if (CellOutOfBounds(head_new))
	{
		g->snake_is_dead = true;
		return;
	}

	for (int i = 1; i < SnakeHead(g); ++i)
	{
		if (Vector2Equals(head_new, g->snake[i]))
		{
			g->snake_is_dead = true;
			return;
		}
	}

	memcpy(g->snake, g->snake + 1, SnakeHead(g) * sizeof(g->snake[0]));
	g->snake[SnakeHead(g)] = head_new;

	if (Vector2Equals(head_new, g->apple))
	{
		g->snake_count++;
		memcpy(g->snake + 1, g->snake, SnakeHead(g) * sizeof(g->snake[0]));
		g->apple = RandomVacantCell(g);
	}
}

void
Update(SnakeGame *g)
{
	if (g->snake_is_dead == true)
	{
		if (IsKeyPressed(KEY_SPACE))
			InitGame(g);
		return;
	}

	Direction pressed_dir = 0;
	for (int key; key = GetKeyPressed();)
	{
		pressed_dir = key;
		if (pressed_dir == SNAKE_LEFT && g->snake_dir != SNAKE_RIGHT)
			g->snake_dir = SNAKE_LEFT;
		if (pressed_dir == SNAKE_RIGHT && g->snake_dir != SNAKE_LEFT)
			g->snake_dir = SNAKE_RIGHT;
		if (pressed_dir == SNAKE_UP && g->snake_dir != SNAKE_DOWN)
			g->snake_dir = SNAKE_UP;
		if (pressed_dir == SNAKE_DOWN && g->snake_dir != SNAKE_UP)
			g->snake_dir = SNAKE_DOWN;
	}

	float time_now = GetTime();
	if ((time_now - g->time_of_last_update) * SNAKE_SPEED >= 1 || pressed_dir != 0)
	{
		g->time_of_last_update = time_now;
		UpdateHead(g);
	}
}

void
Draw(SnakeGame *g)
{
	BeginDrawing();
	{
		ClearBackground(BLACK);

		if (g->snake_is_dead == true)
		{
			int width = MeasureText("GAME OVER", 20);
			DrawText("GAME OVER", (GetScreenWidth() - width) / 2, 40, 20, RAYWHITE);
		}

		for (int i = 0; i < SnakeHead(g); ++i)
			DrawRectangleRec(CellScreenRec(g->snake[i]), LIME);
		DrawRectangleRec(CellScreenRec(g->snake[SnakeHead(g)]), GREEN);

		DrawRectangleRec(CellScreenRec(g->apple), RED);
	}
	EndDrawing();
}

void
UpdateDrawFrame()
{
	Update(&gGame);
	Draw(&gGame);
}

int
main()
{
	InitWindow(GRID_SQUARE_SZ * GRID_W, GRID_SQUARE_SZ * GRID_H, "Snake");
	InitGame(&gGame);

#if defined(PLATFORM_WEB)
	emscripten_set_main_loop(UpdateDrawFrame, 0, 1);
#else
	SetTargetFPS(60);
	while (!WindowShouldClose())
		UpdateDrawFrame();
#endif
	CloseWindow();
	return 0;
}